tests/MongoDB.Bson.Tests/ObjectModel/Decimal128Tests.cs

```diff
@@ -76,14 +74,25 @@ public void Decimal(string valueString, string s)
         [InlineData("1E-57", "0")]
         [InlineData("1E-99", "0")]
         [InlineData("1E-6111", "0")]
-        [InlineData("10000.0000000000000000000000001", "10000")] // see: CSHARP-2001
+        [InlineData("10000.0000000000000000000000001", "10000.000000000000000000000000")] // see: CSHARP-2001
         public void ToDecimal_should_return_expected_result(string valueString, string expectedResultString)
         {
             var subject = Decimal128.Parse(valueString);
             var expectedResult = decimal.Parse(expectedResultString);
 
             var result = Decimal128.ToDecimal(subject);
 
+            result.ToString().Should().Be(expectedResult.ToString());
+        }
+
+        [Theory]
+        [InlineData(0xE000_2000_0000_0000UL, 0)]
```

Comment:
I'm really sorry, but I'm having some understanding how this conversion/rounding works: would you be able to explain how 0XE000_2000_0000_0000 converts to 0?

Reply:
Just a small remark, I've tried to explain the idea of how it works in the first comment of this PR.

> Is the leading bit the sign bit?

yes

> Are we using the "binary integer significand" format or the "densely packed decimal significand" format?

we use ` binary integer significand`

>Would you be able to help me understand how

it happens in this condition: https://github.com/DmitryLukyanov/mongo-csharp-driver/blob/df40385243f7f406196155703a3c34fc4f247ab0/src/MongoDB.Bson/ObjectModel/Decimal128.cs#L775

The idea of this test is the following. The numbers which are in the `second` form are always converted to `0`. The `second form` is the form of 128 bit number which corresponds to next two conditions:
1. The number is conctucted by this template:

>the sign bit (1 bit) with the combination bits (17 bits) with the significand continuation bits (110 bits) for a total of 128 bits

2. The `combination` and `significand continuation` bits are chosen according to the table in this wiki [article](https://en.wikipedia.org/wiki/Decimal128_floating-point_format). See screenshot:
![image](https://user-images.githubusercontent.com/44870738/56811417-533a3f80-6841-11e9-9277-2a630e83a114.png)

Reply:
In our solution(not sure how it happens in other) the 128 bit numbers are built from two 64 bits parts (`highBits` and `lowBits`). 

In this test, `0xE000_2000_0000_0000UL` is the first 64 bits from the whole number which is built based on the first line from wiki table:
    
    |1100mmmmmmmmmmmmm | 00xxxxxxxxxxxx |100c|

See:
0xE000_2000_0000_0000UL = `0b1110_0000_0000_0000_0010_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000`

So, it can be parsed as:
1 (`sign`) 110_0000_0000_0000_00 (`combination bits`)10_00 (`significand msbits`) 00_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000 (`significand continuation bits based on the exponent`)

Since the wiki line which I use to create a number has the exponent value which is started with zeros `00xxxxxxxxxxxx`, I completed significand continuation bits in `highBits` and `lowBits` by zeros for a total of 128 bits.

Reply:
I found this while doing [some research](https://en.wikipedia.org/wiki/Decimal_floating_point):  

> For Decimal128, the second encoding form is actually never used; the largest valid significand of 1034−1 = 1ED09BEAD87C0378D8E63FFFFFFFF16 can be represented in 113 bits.

So perhaps we may be able to away with not testing the second form?

Reply:
For me, if we have some code path, then having at least one test for this part is almost required. Especially for cases like this, when sometimes is difficult to say what is the purpose of some code line/block.
So, I see 2,5 options here:
1. Add tests (as is or similar to). Also we can leave a comment about why this test ahs been written.
2. Remove checking(and tests) on the second form since it's possibly not used, but I believe it's a really bad idea.
2.5. Remove tests and leave a comment in place when we use checking on the `second form` . For me, the approach with tests looks clearer and more expectable.

Reply:
jira formatting :(

Reply:
Also thanks so much for the explanation! I'm still working my way through it, but here are some thoughts I wanted to get down before the week ended:

 - I'm not sure I'm convinced the significant ms bits should be concatenated to the combination bits.
 - If we need to include this test, I may need to Zoom you after doing a bit more reading--perhaps with @rstam  as well I think he implemented the Decimal128 related classes in the driver.
 - If we can get away with not doing this test, perhaps we should skip it as there's a lot of complexity here for uncertain gain. Perhaps @jyemin @rstam could chime in here.

Reply:
As for me, I think testing that branch of the code does seem reasonable but it does seem to be high effort low yield: perhaps we could file a low priority ticket and defer the work till later, although there's also the cost of remembering when back to it later so perhaps we should just bite the bullet and finish the test now.

Reply:
To be honest, I'm not sure that we need to really know how this value has been generated. I checked (in debugging) that this value corresponds with the condition for `second form`, so if we don't have a bug in the code, this number is in `second form` and currently this cased is covered by unit tests.

I don't really see your concerns about why we should not add this test. But if you (or someone else) really don't want to include this test, I don't mind to remove it. 

Sure, we can have a zoom meeting.

Reply:
Oh, I definitely trust you that the value corresponds to the condition in `second form` :) and I also admit this is all new to me and that I have a lot more reading that I need to do.

That being said, I do have another question (that perhaps can be saved for our offline discusion): Perhaps I'm totally reading this incorrectly :)  but: Is `Decimal128.FromIEEEBits` constructing a Decimal128 from the [IEEE-754-2008](https://jira.mongodb.org/secure/attachment/102698/P754-draft-1_5_0.pdf) "binary interchange format" (p17 in the PDF)? If so, isn't the table in the wiki article is showing the binary for the numbers in the decimal128 format and thus should not be used as a test input to this function?

Reply:
>Is Decimal128.FromIEEEBits constructing a Decimal128 from the IEEE-754-2008 "binary interchange format"?

yes, we will have `Decimal128` number in the result.

> isn't the table in the wiki article is showing the binary for the numbers in the decimal128 format 

yes, in this table we see `decimal128` numbers in binary format. See the below comment.

>thus should not be used as a test input to this function?

I guess I see your point, you mean that the input number is already in the `Decimal128` format and no need to convert this input number to `Decimal128` once again? If so, it sounds reasonable.

I'm not really sure in my next comment. @rstam, please correct me if/when I wrong (if any of below makes sense :) ).

As I understand based on this phrase from wiki page:

>IEEE 754 allows two alternative representation methods for decimal128 values. 

the wiki describes the binary formats of `Decimal128` number based on `IEEE 754` standard. So, I assume that the input test number(`0xE000_2000_0000_0000UL`) is `binary decimal128` according to `IEEE754`. To be more precise, this is the first 64 bits of the full number. 
Then, this line:

> var subject = Decimal128.FromIEEEBits(highBits, 0x0);

converts `IEEE` number to our internal `Decimal128` format (the low bits are just zeros). We discussed it with Robert:

>Our internal representation of the 128 bits is *slightly* different from the IEEE one because we wanted a `struct` value of 128 zeros to be equal to 0.

So, then we convert out `internal` `decimal128` value to the .NET decimal here:

>var result = Decimal128.ToDecimal(subject);

Reply:
> > thus should not be used as a test input to this function?
> 
> I guess I see your point, you mean that the input number is already in the `Decimal128` format and no need to convert this input number to `Decimal128` once again? If so, it sounds reasonable.

As discussed offline with @DmitryLukyanov: this is the key issue that we need to figure out. Waiting for @rstam's thoughts: we could definitely all hop onto a Zoom call if a more depth discussion is needed.

Reply:
I'm confused about what you are waiting for.

I think this test is fine (and should have been here to begin with). It is testing that second from Decimal128 is converted to zero (as per the Wikipedia article).

Reply:
1. Just confirming: `Decimal128.FromIEEEBits`'s input are the high bits and low bits of an IEEE `decimal128` (and not their `binary128` format)?

Reply:
Yes, the input to `Decimal128.FromIEEEBits` are the 128 bits of a IEEE Decimal128 in binary integer decimal format, expressed as two 64-bit ulongs.

    /// <summary>
    /// Creates a new Decimal128 value from the IEEE encoding bits.
    /// </summary>
    /// <param name="highBits">The high bits.</param>
    /// <param name="lowBits">The low bits.</param>
    /// <returns>A Decimal128 value.</returns>

I don't know what a `binary128` is.

Reply:
TLDR: I believe our test input is correct and expected output are correct. However, I believe our current interpretation of the input may be incorrect.

Our input is:
> 0xE000_2000_0000_0000UL = `0b1110_0000_0000_0000_0010_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000`

Previous discussions indicate that we're currently thinking of the input like this : 
> So, it can be parsed as:
> 1 (`sign`) 110_0000_0000_0000_00 (`combination bits`)10_00 (`significand msbits`) 00_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000 (`significand continuation bits based on the exponent`)

I believe it is more accurately parsed as

1 (`sign`) 110_0000_0000_0000_00 (`combination bits with the exponent starting after the 11 i.e. the exponent starts in bit 3 of the combination field`)10_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000 (`significand with an implicit leading 100`)

as noted in the [wiki article](https://en.wikipedia.org/wiki/Decimal128_floating-point_format#Binary_integer_significand_field):
> If the 2 bits after the sign bit are "11", then the 14-bit exponent field is shifted 2 bits to the right (after both the sign bit and the "11" bits thereafter), and the represented significand is in the remaining 111 bits. In this case there is an implicit (that is, not stored) leading 3-bit sequence "100" in the true significand.


It makes no difference in that the input still converts to 0, unless we want to ensure that we're testing the edge case of the start of the range for the second form.

Reply:
> Yes, the input to `Decimal128.FromIEEEBits` are the 128 bits of a IEEE Decimal128 in binary integer decimal format, expressed as two 64-bit ulongs.
> 
> ```
> /// <summary>
> /// Creates a new Decimal128 value from the IEEE encoding bits.
> /// </summary>
> /// <param name="highBits">The high bits.</param>
> /// <param name="lowBits">The low bits.</param>
> /// <returns>A Decimal128 value.</returns>
> ```
> 
> I don't know what a `binary128` is.

The IEEE 754-2008 spec referenced in the MongoDB decimal128 spec defines a `binary128` format as well as a `decimal128` format, which is where my confusion regarding the input format came from. Thanks for clearing that up Robert! :)

Reply:
I agree that there is an implied 100 in front of the significand in second from.

But we don't process the significand in second form.

You're not concerned we have a bug are you?

Reply:
Nope, there's no bug.  I'm just concerned about the test coverage/the intent of the test value. Perhaps we could add another test 
`[InlineData(0xE000_0000_0000_0000UL, 0)]` (which passes locally on my machine fwiw) which would translate to `‭1110_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000‬`, which should test the very start of the range for the second form?

Reply:
All second form values convert to zero.

I'm fine with adding more test values at the edges of the range.

Reply:
I've added one more test.

Reply:
Thanks Dima and Robert! 😺